src/hotspot/share/services/heapDumper.cpp

```diff
@@ -2063,7 +2063,6 @@ void DumpMerger::set_error(const char* msg) {
 // read+write combination, which would require transferring data to and from
 // user space.
 void DumpMerger::merge_file(const char* path) {
-  assert(!SafepointSynchronize::is_at_safepoint(), "merging happens outside safepoint");
```

Comment:
This might fix the failure but why is this restriction in place to begin with? This seems like a very expensive operation to be performing by the VMThread during a safepoint!

Reply:
Main idea of segmented heap dump feature (JDK-8306441, #13667) is to divide heap dump process to 2 phases - dump to segment files (this phase is executed at safepoint) and merge of the segment files (this one is supposed to be executed outside of safepoint, so VM is not blocked in this phase). And this asserts were added to ensure it works as desired.
I think merge stage can be performed on the current thread without VMOperation, but I don't know all possible consequences of the change.

Reply:
Prior to the virtual thread changes what thread is performing the merge? After these changes what thread would do it if not included in the VM operation?

The assert is correct - we should not do the merge at the safepoint (or even in the VMThread).

Reply:
> Prior to the virtual thread changes what thread is performing the merge?
Prior the change merge is performed on either AttachListener thread (when it's the current thread - special case for tools like jcmd) or on VMThread (if the current thread is not AttachListener thread).
Note also that prior the change segmented dumping was conditional - if GC supports multiple workers and if the caller requests parallel dump. HeapDumpAfterFullGC/HeapDumpBeforeFullGC request dumping in 1 thread, so there was no merge in the scenarios.

> After these changes what thread would do it if not included in the VM operation?
On the current thread. As I wrote before I 'm not sure if it can cause some unexpected consequences in some cases.
I tried to do the change - tier1 passed (TestReduceAllocationAndHeapDump.java test calls Runtime.gc() with HeapDumpAfterFullGC, merge is performed on main java thread)

Reply:
Sorry, I was wrong. TestReduceAllocationAndHeapDump.java failed on the assert.
I.e. merge is performed on main java thread, but VM is at safepoint

Reply:
> I.e. merge is performed on main java thread, but VM is at safepoint

So is the main thread operating in_native whilst doing the merge? I suspect the admonition of not doing the merge at a safepoint actually meant "not a safepoint by the VMThread" as that would cause the whole VM to pause. Even doing it in the VMThread at all can delay the next safepoint, which does not seem good.

I'm not familiar with this code in general (and only looked at this because of the previous issue in the CI) but I'm unclear what including virtual thread stack referencves has to do with the merging logic? I would not expect merging to be affected by the current change.

Reply:
> So is the main thread operating in_native whilst doing the merge? I suspect the admonition of not doing the merge at a safepoint actually meant "not a safepoint by the VMThread" as that would cause the whole VM to pause. Even doing it in the VMThread at all can delay the next safepoint, which does not seem good.

I think the thread is in_vm.
Main java thread call Runtime.gc(), GC itself calls heap dumper before/after full gc.
I don't know if there is a performance difference in the case between executing merge on the current thread and VMThread.
HeapDumpBeforeFullGC/HeapDumpAfterFullGC/HeapDumpOnOutOfMemoryError are diagnostic flags, so I think performance is not so important here.
There are plans to improve their performance by turning on parallel dump (using GC worker) - often parallel heap dump + merge takes less time than single-threaded dump (we have now).

> I'm not familiar with this code in general (and only looked at this because of the previous issue in the CI) but I'm unclear what including virtual thread stack referencves has to do with the merging logic? I would not expect merging to be affected by the current change.

This is caused by a nature of unmounted virtual threads.
We need to dump stack traces before we dump heap objects, but we can find unmounted virtual threads only by iterating over heap. So we need a way to add data to the beginning of the dump file while dumping heap objects. Segmented dump helps here - we have a designated segment to write stack traces.

Reply:
I tested different scenarios (all I know) of heap dump:
- via attach (jcmd);
- JMX (HotSpotDiagnosticMXBean);
- HeapDumpBeforeFullGC/HeapDumpAfterFullGC;
- HeapDumpOnOutOfMemoryError.
I see no problem with merge on the current thread.
Only HeapDumpBeforeFullGC/HeapDumpAfterFullGC causes heap dump (and merge) at safepoint.
But HeapDumpBeforeFullGC/HeapDumpAfterFullGC/HeapDumpOnOutOfMemoryError depend on GC and require much more testing, I think it's better to integrate this fix as it is and update the way merger is executed by a separate follow-up change (with cleanup of related stuff)

Reply:
Okay. I'm concerned that the merge, whether by the current thread whilst in_vm, or by the VMThread, will either prevent the VM from going to a safepoint in a timely manner, or cause the safepoint to be excessively long. I'm not sure how problems in this area will become visible.